### ISSUE TITLE
`bashly` completions for Bash, Fish, Zsh

### ISSUE THREAD

**Issue by EmilySeville7cfg** [DannyBen/bashly#384]
### Description

It would be nice to have them. :)

**Comment by DannyBen** [comment]
My thoughts on the subject:

1. This is a headache to maintain. 
1. It adds another "ball and chain" on the command line - any change requires changing 3 other scripts.
1. There is no other Ruby Gem that installs completions (to my knowledge) - if there is, I am curious to take a look at it.
1. Continuing the above, it will probably mean that users have to install this completion script manually, which detaches it from the installation (or to have another `bashly install-completions` command, which is another overhead).
1. Writing (and testing) completion scripts is tedious, especially since each shell wants its own format, but not only.
1. Writing simple (read: basic) completion scripts for bash (that can work in other shells if they load bash completions) is relatively easy for me, using my [completely](https://github.com/DannyBen/completely) gem (which is used by bashly itself to let users create completions for their generated script).
1. The `bashly` CLI (as well as most of my other CLIs) intentionally have short and easy to write commands. `bashly g` is the same as `bashly generate` and one keystroke shorter than `bashly g<tab>`.

All that said - I am open to it, if most of the above concerns can be mitigated.

**Comment by EmilySeville7cfg** [comment]
I've heard about [carapace](https://pkg.go.dev/github.com/rsteube/carapace#section-readme) package to generate completions. I just need to play with it a little bit to create completions for shells. :)

**Comment by DannyBen** [comment]
> I've heard about [carapace](https://pkg.go.dev/github.com/rsteube/carapace#section-readme) package to generate completions

Interesting. 

I [see](https://github.com/rsteube/carapace-bin) it also has a [list](https://rsteube.github.io/carapace-bin/completers.html) of CLIs it provides completions to. Perhaps adding bashly to it is easier than other implementation routes.

**Comment by DannyBen** [comment]
@EmilySeville7cfg - since you raise this issue, if you want to test my implementation, it would be nice.

With #293, it is now possible to:

1. Run `bashly completions` and get a completions script to stdout (for manual copying to anywhere)
2. Run `bashly completions --install` and (hopefully) get the script installed in one of the system's [bash completions directories](https://github.com/DannyBen/bashly/blob/6eb004109ff2f8aafb6043d4bb4d536dc8ba90fc/lib/bashly/commands/completions.rb#L56-L61)
3. Download the [bash completions script](https://github.com/DannyBen/bashly/blob/master/lib/bashly/completions/bashly-completions.bash) manually and copy it wherever.

**Comment by EmilySeville7cfg** [comment]
I'll check it out today. :)

**Comment by DannyBen** [comment]
All the patterns that can be completed can be easily seen in the [completely.yaml](https://github.com/DannyBen/bashly/blob/master/lib/bashly/completions/completely.yaml) file.

**Comment by DannyBen** [comment]
[Refactored](https://github.com/DannyBen/bashly/blob/master/lib/bashly/commands/completions.rb) completions installation, thanks to the new [`Completely::Installer`](https://github.com/DannyBen/completely/blob/master/lib/completely/installer.rb) class introduced in https://github.com/DannyBen/completely/pull/51

**Comment by EmilySeville7cfg** [comment]
Trying to built the latest version of bashly, `gem install bashly.gem` fails with:

```
Fetching requires-1.0.1.gem
Fetching docopt_ng-0.7.1.gem
Fetching colsole-1.0.0.gem
Fetching mister_bin-0.7.6.gem
Fetching lp-0.2.1.gem
Fetching gtx-0.1.0.gem
Fetching module_methods-0.1.0.gem
Fetching filewatcher-2.1.0.gem
Fetching completely-0.6.0.gem
ERROR:  While executing gem ... (Gem::FilePermissionError)
```

output. `sudo` solves this issue. Maybe add some note about it?

**Comment by EmilySeville7cfg** [comment]
@DannyBen, I am not sure but I would like to suggest to add `bashly completions --uninstall` option too.

**Comment by EmilySeville7cfg** [comment]
After installation and session restart I have no working completions for bashly. I've tried run bash as: `bash` and `bash --norc` (as I use BLE).

**Comment by EmilySeville7cfg** [comment]
Solution: install `bash-completion` package. ;)

**Comment by DannyBen** [comment]
> sudo solves this issue. Maybe add some note about it?

Gems should normally not be installed with sudo

> Solution: install bash-completion package

On non-bash shells, this is needed, yes.
What did you run? `brew install bash-completions` or `apt install bash-completions`?

> `bashly completions --uninstall`

Yes. But this will require an update to completely as well.

**Comment by EmilySeville7cfg** [comment]
> What did you run? brew install bash-completions or apt install bash-completions?

The later one.

**Comment by DannyBen** [comment]
> So... is it my personal issue or not? I use Linux 22.04 Ubuntu.

It is weird that you needed `bash-completions` then. I am also on Ubuntu.
You are probably using the system ruby, which is not ideal.

1. People who use Ruby regularly, install Ruby with a version manager. I am using rbenv and installing my Ruby with [`rush get ruby`](https://github.com/DannyBen/rush-repo/blob/master/ruby/main)
2. People who do not use Ruby regularly, should probably use the bashly docker image, or take the time to set up a proper ruby environment.

A quick search of your error comes up with a [StackOverflow question](https://stackoverflow.com/questions/18599889/error-while-executing-gem-gemfilepermissionerror), with answers to install a version manager.

All that said - it would help if you tell me after your trial-and-errors - what you believe other users should be aware of, if there is anything other than `apt install bash-completions`.

**Comment by EmilySeville7cfg** [comment]
> You are probably using the system ruby, which is not ideal.

I've [installed](https://www.ruby-lang.org/en/documentation/installation/) it via `ruby-full` package. 🤔

**Comment by DannyBen** [comment]
[Released in 1.0.6](https://github.com/DannyBen/bashly/releases/tag/v1.0.6).
